### PR TITLE
ipq40xx: Add support for TP-Link Deco M5 EU v3.0

### PR DESCRIPTION
**Pull request by GoliathLabs**
TP-Link Deco M5 EU (v3.0) is a wireless router

**Specification**
SoC: Qualcomm IPQ4019
RAM: 256 MiB
Flash: 32 MiB
WiFi: Qualcomm IPQ4019 (2,4 GHz + 5 GHz) 802.11bgnac
Ethernet : 1x (10/100/1000 Mbps) IPQ4019
Switch: Qualcomm Atheros QCA8072
LEDs: 1x
USB: 1x 2.0
FCCID: TE7M5V3

**Installation**

It is possible to install [OpenWrt](https://openwrt.org/docs/guide-user/installation/generic.flashing.tftp) on this model. The same method works for restoring the default firmware.

→ generic.flashing.tftp

Connect one of the LAN ports (might need to be specifically port 1) to a machine running a TFTP server. Configure the server machine with IP 192.168.0.66/24 [1]. The router will run as a TFTP client at 192.168.0.88.

Obtain a stock firmware or a 'factory' OpenWrt image. Save the firmware to the TFTP server as M5v3_tp_recovery.bin. Replace the v3 with the version number of your hardware if you don't have a v3.

To start the TFTP recovery process on the router, press and hold the WPS/Reset button and turn on the router. Hold down the WPS/Reset button until the TFTP transfer starts.

If everything went correctly, you should see a read request in your TFTP server. The flash process will take about 3 to 5 minutes.

Further sysupgrades can be performed via the web interface.

Signed-off-by: Felix Golatofski git@xdfr.de

### CODE REVIEW

**Comment by GoliathLabs** [comment]
This is my first PR regarding OpenWrt, so don't expect everything to be perfect. Nonetheless, I hope to bring official support for this device upstream.

The repository I took changes from also modified the safeloader quite a bit. I don't know to what extent this is necessary in this case (https://github.com/lntgn/openwrt/blob/master/tools/firmware-utils/src/tplink-safeloader.c). The maintainer of the repository has contacted me to see if I am willing to add support for this device to the upstream, he does not want credit for it.


In the meantime, you are doing a great job with maintaining this heavy codebase and supporting some older router models for years to come! Thank you very much 😄.

**Comment by lntgn** [comment]
> The repository I took changes from also modified the safeloader quite a bit. I don't know to what extent this is necessary in this case (https://github.com/lntgn/openwrt/blob/master/tools/firmware-utils/src/tplink-safeloader.c).

Instead of fixing the conflicts that came up when merging from frankveltmans/openwrt I just copied the file verbatim. It worked well enough for my own use, but it certainly wasn't appropriate to submit a pull request with that bastardized tplink-safeloader.c

Thank you for taking the time to do it correctly.

**Comment by GoliathLabs** [comment]
No problem, didn't take that long. Just do a quick grep for "Deco M5" and you have the appropriate sections to move to the master's safeloader.

**Comment by GoliathLabs** [comment]
This could also work for the non-EU variant, as the only differences seem to be in the power supply.

**Comment by GoliathLabs** [comment]
If a reviewer says things are good as they are now, I could change things to an international format I guess? (Deco M5 v3)

Advice from an advanced contributor would be appreciated here 😄

**Comment by lntgn** [comment]
I just tried building this, your tplink-safeloader.c spits out a bunch of errors. It needs to be reworked it into a format compatible with the current state of openwrt. That's where I got stuck.

**Comment by GoliathLabs** [comment]
Can you provide me the log?

**Comment by GoliathLabs** [comment]
I might have a look at it later

**Comment by lntgn** [comment]
> Can you provide me the log?

https://pastebin.com/UY55YEUU

Thank you.

**Comment by adschm** [comment]
This has formal issues.

Please clean up your commit message and add Specifications, Flashing instructions, and MAC addr table.

https://openwrt.org/submitting-patches
https://openwrt.org/docs/guide-developer/device-support-policies

File: package/firmware/ipq-wifi/Makefile
Comment (adschm) [inline package/firmware/ipq-wifi/Makefile:54]: Typically, TP-Link uses the same hardware for all y in vx.y.

So, would v3 be enough here instead of v3.0?

Do you know whether non-eu versions are different? eu should only be added if it's different from the other v3, and when this is the case it should be last, i.e. Vendor Model Vx (EU) and vendor_model-vx-eu ...

File: target/linux/ipq40xx/base-files/etc/board.d/02_network
Comment (adschm) [inline target/linux/ipq40xx/base-files/etc/board.d/02_network:120]: Vendor prefix for TP-Link is tplink in the compatible. You can make your and our job much easier by having a look at what other people have done in similar cases before you ...

Comment (adschm) [reply target/linux/ipq40xx/base-files/etc/board.d/02_network:120]: And for the particular setup here, this can be merged with the proper case above.

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
Comment (adschm) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:164]: Please use the proper partition name. use "2" instead of "+2".

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
Comment (adschm) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:282]: See above.

File: target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh
Comment (adschm) [inline target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh:33]: Alphabetic sorting.

File: target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh
Comment (adschm) [inline target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh:36]: See above. But this looks like it can be set in DTS instead and removed here anyway.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts:14]: use SPDX instead.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts:1]: This is already part of the kernel?

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts:41]: reset

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts:42]: "reset"

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts:52]: Normally, it's color:function. What's the function of these LEDs?

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts:155]: flash@0

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts:149]: DT label can be removed.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts:171]: numbering before @ should be removed, just partition@0, partition@30000 etc.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts:318]: Broken indent here and below.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts:360]: missing newline at EOF.

File: target/linux/ipq40xx/image/generic.mk
Comment (adschm) [inline target/linux/ipq40xx/image/generic.mk:762]: This line can be dropped.

File: target/linux/ipq40xx/image/generic.mk
Comment (adschm) [inline target/linux/ipq40xx/image/generic.mk:756]: See my earlier comment on naming.

File: target/linux/ipq40xx/image/generic.mk
Comment (adschm) [inline target/linux/ipq40xx/image/generic.mk:759]: EU is part of variant, and probably there should be a space before the M5 (at least it is in your PR title) ...

File: target/linux/ipq40xx/image/generic.mk
Comment (adschm) [inline target/linux/ipq40xx/image/generic.mk:764]: BOARD_NAME can be removed.

File: target/linux/ipq40xx/image/generic.mk
Comment (adschm) [inline target/linux/ipq40xx/image/generic.mk:767]: consider IMAGES += factory.bin

File: target/linux/ipq40xx/patches-5.4/901-arm-boot-add-dts-files.patch
Comment (adschm) [inline target/linux/ipq40xx/patches-5.4/901-arm-boot-add-dts-files.patch:51]: This looks unrelated.

File: tools/firmware-utils/src/tplink-safeloader.c
Comment (adschm) [inline tools/firmware-utils/src/tplink-safeloader.c:2477]: Sorting ...

File: tools/firmware-utils/src/tplink-safeloader.c
Comment (adschm) [inline tools/firmware-utils/src/tplink-safeloader.c:2509]: Ah, so now you are contradicting yourself, since this is neither limited to 3.0 nor to EU.
And please do not add 1.0 strings to a v3. This should be sorted, and please check whether the partition table in all of the devices finally chosen is the same.

File: tools/firmware-utils/src/tplink-safeloader.c
Comment (adschm) [inline tools/firmware-utils/src/tplink-safeloader.c:2533]: Is this @0 syntax correct here?

File: tools/firmware-utils/src/tplink-safeloader.c
Comment (adschm) [inline tools/firmware-utils/src/tplink-safeloader.c:3000]: What's that supposed to do. This should be discussed in your commit message. If you have to introduce that much logic, I'd even consider to make the tplink-safeloader changes a separate commit before adding the actual device.

**Review by adschm — CHANGES_REQUESTED** [review]

File: package/firmware/ipq-wifi/Makefile
Comment (lntgn) [reply package/firmware/ipq-wifi/Makefile:54]: I did not submit the PR but I will reply since it may help with this effort.

Yes, v3 would do the trick. V3.2 and V3 have identical hardware. I believe V1 also, since TP-Link reuses the same firmware for V1, V3 and V3.2 variants. Those have NOR flash, while the V2 model has NAND flash. (Not sure if there are other differences)

International variants only differ in the provided power supply. The CA and US devices have a 120V adapter along with a barrel connector, the EU variant I believe has a USB-C port for power (and only power, not hooked up to USB logic).

File: tools/firmware-utils/src/tplink-safeloader.c
Comment (lntgn) [reply tools/firmware-utils/src/tplink-safeloader.c:2509]: Some more info about this device... V1, V3 and V3.2 seem compatible with each other. I cannot say for 100% certainty, but the stock firmware is identical for all three. It only differs on V2... with that said, I have a V3 / CA working flawlessly. These "special_id's" match what is found in stock firmware.

File: target/linux/ipq40xx/patches-5.4/901-arm-boot-add-dts-files.patch
Comment (lntgn) [reply target/linux/ipq40xx/patches-5.4/901-arm-boot-add-dts-files.patch:51]: AFAIK they are unrelated. Since it's a patch file, github is showing green lines where there shouldn't be... the only line that was actually added is line 48. 49, 50 and 51 just moved down one line.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts
Comment (lntgn) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-deco-m5-eu-v3.0.dts:52]: I am posting this in case it helps - there is only 1 LED on the device, it is an RGB LED and each color. I don't believe there is any pwm.

File: package/firmware/ipq-wifi/Makefile
Comment (adschm) [reply package/firmware/ipq-wifi/Makefile:54]: v1 and v3 should be separate then, and this should probably be named v3 without region ...

File: tools/firmware-utils/src/tplink-safeloader.c
Comment (adschm) [reply tools/firmware-utils/src/tplink-safeloader.c:2509]: v1 and v3 should be treated separately if v2 is different.

File: target/linux/ipq40xx/patches-5.4/901-arm-boot-add-dts-files.patch
Comment (adschm) [reply target/linux/ipq40xx/patches-5.4/901-arm-boot-add-dts-files.patch:51]: I don't think so, these lines are duplicated.

**Comment by GoliathLabs** [comment]
@lntgn I have invited you to the repository. I'm in the middle of my exams at university right now, so I don't have that much time to sort things out. Maybe you want to try it out

**Comment by lntgn** [comment]
Okay, I am waiting for more devices to come in the mail. They are V2 US, I will compare them to my V3 CA and if they are the same we should be able to remove EU_V3 and support all Deco M5's. Thanks for your help Felix

On July 2, 2021 9:35:49 a.m. ADT, Felix ***@***.***> wrote:
***@***.*** I have invited you to the repository. I'm in the middle of my exams at university right now, so I don't have that much time to sort things out. Maybe you want to try it out

**Comment by wulfy23** [comment]
when the commit message is sorted... please clarify usb port type and feature-set...

**Comment by gustavargas** [comment]
Reporting from:

SoC: Qualcomm IPQ4019
RAM: 256 MiB
Flash: 32 MiB
WiFi: Qualcomm IPQ4019 (2,4 GHz + 5 GHz) 802.11bgnac
Ethernet : 1x (10/100/1000 Mbps) IPQ4019
Switch: Qualcomm Atheros QCA8072
LEDs: 1x
USB: 0x 2.0
FCCID: TE7M5V3
Version compiled: OpenWrt SNAPSHOT r17048-f32b247c50 / LuCI Master git-21.196.22411-7d9ab8d
Variant: US/3.0 with power cable, no usb port.

Working without any problem. Flashed with TFTP method.

**Comment by Gamerayers** [comment]
Any chance it works on the Deco M9 also?  I know the major difference between M5 and M9 is that the M9 has additional hardware for Home Automation Hub?

**Comment by robhanlon22** [comment]
@gustavargas did you have to make any modifications to get it working on your US/3.0?

**Comment by gustavargas** [comment]
> @gustavargas did you have to make any modifications to get it working on your US/3.0?

No, I just compile the source from the pull request (https://github.com/MorphNetwork/openwrt) and flashed with the TFTP method. 3 days working just fine in 3 M5 US 3.0 Version.

**Comment by LennardF1989** [comment]
Does the whole mesh feature still work with this? Or will it become a "normal" Router/AP depending on configuration?

**Comment by gustavargas** [comment]
> 
> 
> Does the whole mesh feature still work with this? Or will it become a "normal" Router/AP depending on configuration?

The OpenWRT recognizes the mesh feature, but I didn't use it. I connected with lan cables the APs and works nicely with fast roaming.

**Comment by beandj** [comment]
@gustavargas, How did you build? I get error on tplink-safeloader.c during build. Can you share the compiled .bin? Or explain which selections in make menuconfig ?

**Comment by GoliathLabs** [comment]
@gustavargas or others interested in getting upstream support for this device, feel free to send me pull requests and I will merge them here. I don't have much time right now, so I can't do much work on this at the moment. Help is always appreciated :).

**Comment by adschm** [comment]
Looks like nothing has changed here?

**Comment by lntgn** [comment]
Maybe in the future or if someone wants to pick it up. Babies will do that with your free time lol it evaporates.. good luck


________________________________

**Comment by colinfindlay-nz** [comment]
@lntgn @GoliathLabs 
I might have a few days free - was it only cosmetic and documentation changes required to get this merged?

**Comment by GoliathLabs** [comment]
@colinfindlay-nz There have been some problems with the safeloader and it looks like as if the patch @lntgn and I worked out does work on other Deco's as well which needs to be integrated here. But any contribution would help I guess :)On Dec 17, 2021 22:18, colinfindlay-nz ***@***.***> wrote:
@lntgn @GoliathLabs
I might have a few days free - was it only cosmetic and documentation changes required to get this merged?

**Pull request closed without merging**
